Making TPRs list-able with '--all-namespaces'

arschles opened the pull request:
See the comments in https://github.com/kubernetes-incubator/service-catalog/issues/672#issuecomment-294970242 for details on the problem and this solution.

Requires #715 
Fixes #672

arschles commented:
Now, all of the commands listed in #672 work:

```console
ENG000656:service-catalog aaronschlesinger$ k get instances ups-instance -n test-ns -o yaml
apiVersion: servicecatalog.k8s.io/v1alpha1
kind: Instance
metadata:
  creationTimestamp: 2017-04-18T20:11:51Z
  finalizers:
  - kubernetes
  name: ups-instance
  namespace: test-ns
  resourceVersion: "178021"
  selfLink: /apis/servicecatalog.k8s.io/v1alpha1/namespaces/test-ns/instances/ups-instance
  uid: 4285b6a4-2473-11e7-9609-0800277cda4e
spec:
  osbGuid: 614e812b-519a-459e-a530-e481fa044243
  planName: default
  serviceClassName: user-provided-service
status:
  conditions:
  - lastTransitionTime: 2017-04-18T21:14:47Z
    message: The instance references a ServiceClass that does not exist. Instance
      "test-ns/ups-instance" references a non-existent ServiceClass "user-provided-service"
    reason: ReferencesNonexistentServiceClass
    status: "False"
    type: Ready
ENG000656:service-catalog aaronschlesinger$ k get instances --all-namespaces
NAMESPACE   NAME           KIND
test-ns     ups-instance   Instance.v1alpha1.servicecatalog.k8s.io
ENG000656:service-catalog aaronschlesinger$ k get Instance.v1alpha1.servicecatalog.k8s.io --all-namespaces
NAMESPACE   NAME           KIND
test-ns     ups-instance   Instance.v1alpha1.servicecatalog.k8s.io
```

... when previously only 2/3 of them did.

Note that when a namespaced resource (i.e. an `Instance`) is requested in the same namespace as the server's default TPR namespace, the server behaves in the same way as if `--all-namespaces` was passed to `kubectl`

vaikas-google commented on pkg/storage/tpr/storage_interface.go:
```diff
@@ -325,31 +325,52 @@ func (t *store) List(
-		glog.Errorf("doing request (%s)", err)
-		return err
+	if t.hasNamespace && ns == t.defaultNamespace {
+		// if the resource is supposed to have a namespace, and the given one is the default,
```
Is this the behaviour that current objects do?

arschles replied:
do you mean built-in objects?

vaikas-google replied:
yes, that's what I meant, sorry for not being clearer :)

arschles replied:
got it. no, this isn't the behavior of built-in objects, because they don't have the concept of a "default namespace" (because they are stored in etcd).

vaikas-google commented:
Looks like travis failed due to gofmt failing on:
pkg/storage/tpr/list_resources_test.go

krancour commented:
@arschles... I think you're afk today... I moved this back into an in-progress status as I've had some important revelations about unit testing the tpr-backed storage impl that I think you and I should discuss before we further consider merging either this or #715-- both of which add new unit tests for the tpr-backed storage.

pmorie commented:
@arschles looks like a gofmt error

krancour commented:
@arschles following up on my previous comment... I have discussed with @pmorie to confirm my revelation: Our unit tests for tpr-backed storage should be using internal/unversioned types as the inputs and outputs. (irl, the registries are the clients of the storage layer and they wheel and deal in internal/unversioned types.)

arschles commented:
@vaikas-google @pmorie the `gofmt` issue is now fixed

@krancour now that #715 is in, I rebased this onto that, and I merged https://github.com/arschles/kubernetes-incubator-service-catalog/pull/3 into this branch, it looks like this is using the proper types. Thanks for pointing that out

krancour commented on pkg/storage/tpr/list_resources_test.go:
```diff
@@ -0,0 +1,117 @@
+		ObjectMeta: metav1.ObjectMeta{Name: "broker2"},
+	}
+	cl.storage[ns] = newTypedStorage()
+	cl.storage[ns][ServiceBrokerKind.URLName()] = objStorage
```
I like this strategy of manipulating the fake's core apiserver client's underlying in-memory storage directly to set up the test. I will adopt that in a follow-up to #715. 👍

arschles commented on pkg/storage/tpr/init_test.go:
```diff
@@ -0,0 +1,50 @@
+)
+
+func serviceCatalogAPIGroup() testapi.TestGroup {
+	// OOPS: didn't register the right group version
```
@pmorie I'll go do a follow up right now

arschles replied:
@pmorie see https://github.com/kubernetes-incubator/service-catalog/pull/729
